lib/Sema/CSDiagnostics.cpp

```diff
@@ -110,18 +110,9 @@ Expr *FailureDiagnostic::findParentExpr(const Expr *subExpr) const {
   return cs.getParentExpr(const_cast<Expr *>(subExpr));
 }
 
-Expr *
-FailureDiagnostic::getArgumentListExprFor(ConstraintLocator *locator) const {
-  auto path = locator->getPath();
-  auto iter = path.begin();
-  if (!locator->findFirst<LocatorPathElt::ApplyArgument>(iter))
-    return nullptr;
-
-  // Form a new locator that ends at the ApplyArgument element, then simplify
-  // to get the argument list.
-  auto newPath = ArrayRef<LocatorPathElt>(path.begin(), iter + 1);
-  auto argListLoc = getConstraintLocator(locator->getAnchor(), newPath);
-  return getAsExpr(simplifyLocatorToAnchor(argListLoc));
+ArgumentList *
+FailureDiagnostic::getArgumentListFor(ConstraintLocator *locator) const {
+  return getConstraintSystem().getArgumentList(locator);
```

Comment:
See my comment about `ArgumentLists` - It'd be great if we could extract this information for `Solution` instead because we can't reply on constraint system having solution applied always e.g. ambiguity cases.